File: src/Master.c
/* Course: COMP-4320 Introduction to Networks
 * Institution: Auburn University
 * Submitting: Laboratory Assignment 2
 * Date: October 25, 2018
 * Authors: Paul Ryu, William Atkinson, Abriana Fornis
 * Emails: phc0004@auburn.edu, wja0007@auburn.edu, adf0018@auburn.edu
 * Resources used: JetBrains CLion IDE, Github, Sublime Text 3, Visual Studio Code
 *
 * Description: The objective of this lab is to create a virtual ring of nodes over the Internet. After joining
 *             the ring, the nodes on the ring MUST use only the ring to communicate. The ring is managed by a
 *             master which is part of the ring and has Ring ID 0 (zero). All the other nodes of the ring are slave
 *             nodes (clients) with a ring ID assigned by the master.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h> 
#include <sys/types.h> 
#include <sys/socket.h> 
#include <netinet/in.h> 
#include <netdb.h> 
#include <arpa/inet.h> 
#include <sys/wait.h> 
#include <signal.h> 

#define PORTNUM "10120" // needs to change for command line function
#define BACKLOG 10

int main(int argc, char *argv[]) {

    int ringID = 0; // ring ID for master
    int nextRID = 1; // assign & increment whenever a node joins
    int myGID = 22;
    int myRID = 0; // ring ID for master
    int slaveRID = 1; // assign & increment whenever a node joins
    int nextSlaveIP = 0;  //Must be initially set to machine IP address, changed to IP of joining slave later
    int nextSlaveID = 0;

    unsigned const int magic_number = 0x4A6F7921;
    const char* portNum;
    int packed_clientIP = -1;

    socklen_t addr_size;
    struct addrinfo hints, *result;
    struct sockaddr_in slave_info, *nextSlaveIP;
    int sockfd, new_fd;

    // check command line arguments
    if (argc == 2) {
        portNum = argv[1];
    }
    else if (argc < 2) {
        printf("Error: please provide the port number\n!");
        exit(1);
    }
    else {
        printf("Error: too many arguments\n!");
        exit(1);
    }

    // load address structs with getaddrinfo()
    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_PASSIVE; // fills in my IP address automatically

    int status;
    if (status = getaddrinfo(NULL, portNum, &hints, &result) != 0) {
        printf("Something went wrong with call to getaddrinfo(): \n", gai_strerror(status));
        exit(1);
    }
    // assign Master IP address to nextSlaveIP
    
    // create socket, bind it, and listen on it
    sockfd = socket(result->ai_family, result->ai_socktype, result->ai_protocol);
    printf("Socket created.\n");

    if (bind(sockfd, result->ai_addr, result->ai_addrlen) < 0) {
        printf("Error: bind failed!");
        exit(1);
    }

    printf("Master: bound to socket\n");
    if (listen(sockfd, BACKLOG) < 0) {
        printf("Error: Something went wrong listening on port ", portNum);
        exit(1);
    } else {
        printf("Master: listening on port %s", portNum, "\n");
    }

    // accepting incoming connections
    while (true) {
        printf("Master: waiting for a connection");
        addr_size = sizeof slave_info;
        new_fd = accept(sockfd, (struct sockaddr *)&slave_info, &addr_size);
        if (new_fd < 0) {
            printf("Error: Could not connect to client!");
            exit(1);
        }
        printf("Master: received connection from %s", inet_ntoa(slave_info.sin_addr));
    }

    freeaddrinfo(result); 
    return 0;
}

File: src/Master.java
/* Course: COMP-4320 Introduction to Networks
 * Institution: Auburn University
 * Submitting: Laboratory Assignment 2
 * Date: October 25, 2018
 * Authors: Paul Ryu, William Atkinson, Abriana Fornis
 * Emails: phc0004@auburn.edu, wja0007@auburn.edu, adf0018@auburn.edu
 * Resources used: JetBrains CLion IDE, Github, Sublime Text 3, Visual Studio Code
 *
 * Description: The objective of this lab is to create a virtual ring of nodes over the Internet. After joining
 *             the ring, the nodes on the ring MUST use only the ring to communicate. The ring is managed by a
 *             master which is part of the ring and has Ring ID 0 (zero). All the other nodes of the ring are slave
 *             nodes (clients) with a ring ID assigned by the master.
 *
 * Master Port Number = (Group ID % 30) * 5 + 10010
 *      Step 1: 22 % 30 = 22
 *      Step 2: 22 * 5  = 110
 *      Step 3: 110 + 10010 = 10120
 *      Result: 10120
 *
 *
 * Compilation Instructions on the Tux Machines for This File (Group 22)
 *      Step 1: javac Master.java
 *      Step 2: java Master 10120
 *
 */

//---------------------------------- Imports ----------------------------------
import java.io.*;
import java.net.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.net.InetAddress;
import java.net.ServerSocket;

/*
//------------------------------ Main Master Class ----------------------------
// Class name: Master
// Function: Main class, acts as the entry point for code and functions for Master.
// Main variables: GID, RID, Port, SlaveIP
*/
public class Master {

	/*
    Method name: main
    Function: main() function, entry point for arguments put into command line
    Behaviors: Open and close socket
    */
    public static void main(String[] args) throws IOException {

        //------------------------------- Variables -------------------------------

        // Initialize NEXT_RING_ID
        byte NEXT_RING_ID = 0;

        // A byte array sent as a byte array in Slave.py
        byte[] requestFromSlave;

        /*
        https://stackoverflow.com/questions/9481865/getting-the
            -ip-address-of-the-current-machine-using-java
            This obtains the IP of the current machine.
        */
        byte[] NEXT_SLAVE_IP = InetAddress.getLocalHost().getAddress();

	    // Only the Master Port Number should be put into the command line.
		if (args.length != 1) {
			throw new IllegalArgumentException("Invalid port number or too many arguments " +
                                               "entered. Please try again.");
		}

		// Convert the Port Number entered in: java Master 10120
		int masterPortNumber = Integer.parseInt(args[0]);

		// Create new socket in Java
		ServerSocket masterSocket = new ServerSocket(masterPortNumber);

        /* Loop infinitely until new Slaves (Nodes) send join requests for the ring. */
		while(true) {

		    // Accept Slave.py
            Socket sock = masterSocket.accept();

            // Get the NEXT_SLAVE_IP of Slave.py
            byte[] incoming_NEXT_SLAVE_IP = sock.getInetAddress().getAddress();

            // Get the IP address of Slave.py
            String incomingHostAddress = sock.getInetAddress().getHostAddress();

            // Get the port number of Slave.py
            int incomingPortNumber = sock.getPort();

            // Initialize a bytearray to be read from the Slave.py join request.
            requestFromSlave = new byte[32];

            /*
            Organize print statements to display to the screen the
                IP address of Slave.py and the Port Number of Slave.py
            */
            System.out.println(" ------------------------------------------------ ");
            System.out.println("\nConnected to IP Address: " + incomingHostAddress
                    + "\n"
                    + "Port Number: " + incomingPortNumber + "\n");

            // Create a bytearray or message to send back to Slave.py
			byte[] packedMessage = new byte[10];

			// Pack the magic number in four parts due to it being hexadecimal.
			packedMessage[1] = 0x4A;
			packedMessage[2] = 0x6F;
			packedMessage[3] = 0x79;
			packedMessage[4] = 0x21;

			// Handle getting input from Slave.py
			InputStream input = sock.getInputStream();

			// Setup ability to give output to Slave.py
			OutputStream output = sock.getOutputStream();

			// Get input from Slave.py in a bytearray, size 32.
			input.read(requestFromSlave);

			// Display the magic number from the bytearray
			System.out.println("Magic Number: ");
			System.out.print(requestFromSlave[0] + " ");

			// Loop until the complete Magic Number bytes are attained.
			for (int i = 1; i < 5; i++) {
				System.out.print(Integer.toHexString((int)requestFromSlave[i]) + " ");
				} System.out.println("\n");

			// Increment the number of slaves since a slave has been added to the ring.
			NEXT_RING_ID++;

			/*
            Pack the message to send back to Slave.py
            The Group ID are the first few bytes.
            */
            byte GROUP_ID = requestFromSlave[0];
			packedMessage[0] = GROUP_ID;

			// Insert the RID and Slave IP AFTER the magic number.
			packedMessage[5] = NEXT_RING_ID;
			packedMessage[6] = NEXT_SLAVE_IP[0];
			packedMessage[7] = NEXT_SLAVE_IP[1];
			packedMessage[8] = NEXT_SLAVE_IP[2];
			packedMessage[9] = NEXT_SLAVE_IP[3];

			// Join the nodes
			NEXT_SLAVE_IP = incoming_NEXT_SLAVE_IP;

            // Complete, ready to close socket.
			output.write(packedMessage, 0, 10);

			/*
            Display to the user that the slave was successfully added to the ring
                with the index counter.
            */
            System.out.println("Slave " + NEXT_RING_ID + " attached to node ring.\n");

			// Close socket.
			sock.close();
		}
	}
}
